pynicotine/gtkgui/userbrowse.py

```diff
@@ -687,24 +789,233 @@ def on_select_dir(self, selection):
+
+        self.FolderTreeView.collapse_row(path)  # show 2nd level
         self.FolderTreeView.expand_row(path, False)
+
         return True
 
-    def on_collapse_row_accelerator(self, *args):
-        """ Left: expand row """
+    def on_folder_expand_sub_accelerator(self, *args):
+        """ =equal: expand only (dont move focus)   """
 
         path, _focus_column = self.FolderTreeView.get_cursor()
-        self.FolderTreeView.collapse_row(path)
+        self.FolderTreeView.expand_row(path, False)
+        return True
+
+    def on_folder_focus_filetree_accelerator(self, *args):
+        """ Shift+Enter: focus selection over FileTree  """
+
+        if len(self.file_store) >= 1:
+            self.FileTreeView.grab_focus()
+
+        else:
+            self.on_folder_expand_sub_accelerator()
+            # ToDo: Expand all branch sub-nodes (*)                                                             # ToDo #
+
+        return True
+
+    def on_folder_transfer_to_accelerator(self, *args):
+        """ Ctrl+Enter: Upload Folder To...
+                        Download Folder Into...         """
+
+        if self.user == config.sections["server"]["login"]:
+            if len(self.file_store) >= 1:
+                self.on_upload_directory_to()
+            else:
+                self.on_upload_directory_recursive_to()
+
+        else:
+            if len(self.file_store) >= 1:
+                self.on_download_directory_to()
+            else:
+                self.on_download_directory_recursive_to()
+
+        return True
+
+    def on_folder_transfer_accelerator(self, *args):
+        """ Shift+Ctrl+Enter: Upload Folder Recursive To...
+            (without prompt)  Download Folder           """
+
+        if self.user == config.sections["server"]["login"]:
+            self.on_folder_expand_sub_accelerator()
+            self.on_upload_directory_recursive_to()
+
+        else:
+            if len(self.file_store) <= 0:
+                # avoid accidental recursive download
+                self.on_folder_expand_sub_accelerator()
+
+            else:
+                self.on_download_directory()  # without prompt
+
+        return True
+
+    def on_folder_open_manager_accelerator(self, *args):
+        """ Ctrl+Alt+Enter: Open folder in File Manager...
+            (or Alt+Enter)  Download Folder Into...    """
+
+        if len(self.file_store) <= 0:
+            self.on_folder_expand_sub_accelerator()
+
+        if self.user == config.sections["server"]["login"]:
+            self.on_file_manager()
+
+        elif len(self.file_store) >= 1:  # [user is not self]
+            self.on_download_directory_to()
+
+        return True
+
+    """ Key Bindings (FileTreeView) """
+
+    def on_file_focus_folder_left_accelerator(self, *args):
+        """ Left: focus back parent folder (left arrow) """
+
+        # ToDo: Only if... FileTreeView.selected_column == 0:  # ToDo: (scrolled fully left)                    # ToDo #
```

Comment:
This functionality is scuppered by an issue with horizontal scrolling. A mouse-less user would be able to scroll right, but not left, meaning that the FileTreeView can get stuck in an unreadable position.

I need to detect which column is selected (or the value of the h-scrollbar), then only allow the focus shift if it's column[0] (left-most), but I'm unaware of how to implement the GtkTreeView methods. I want to do something like:

```
        _path, focus_column = self.FileTreeView.get_cursor()
        first_column = self.FileTreeView.get_column(0)

        if focus_column == first_column:
            self.FolderTreeView.grab_focus()
```
Do you know of a solution for this? Alternatively, now that the userbrowse.ui is reponsive, I see little benefit to horizontal scrolling. Perhaps it would be better to auto-expand the Filename column, but I don't know how to accomplish this either.

Reply:
I didn't think I was going to be able to find a solution for this, but I found one in the end:

```
    def on_file_focus_folder_left_accelerator(self, *args):
        """ Left: focus back parent folder (left arrow) """

        h = self.FileScrolledWindow.get_hadjustment()

        if h.get_value() > 0.0:
            # Allow horizontal scrolling
            return False

        self.FolderTreeView.grab_focus()

        return True
```